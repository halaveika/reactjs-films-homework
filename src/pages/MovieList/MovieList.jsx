import React from 'react';
import { Layout } from 'antd';
import PropTypes from 'prop-types';
import MovieItem from '../../components/MovieItem';
import './MovieList.scss';

export default function MovieList({
  items, video, handleVideo, getDetails,
}) {
  const itemList = items.filter((item) => item.poster && !item.poster.includes('null')).slice(0, 15).map((item) => (
    <MovieItem
      key={item.id}
      id={item.id}
      title={item.title}
      genres={item.genres}
      vote_average={item.vote_average}
      poster={item.poster}
      overview={item.overview}
      video={video}
      handleVideo={handleVideo}
      getDetails={getDetails}
    />
  ));

  return (
    <Layout className="movieList-container">
      {itemList}
    </Layout>
  );
}

MovieList.propTypes = {
  items: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired,
    title: PropTypes.string,
    genres: PropTypes.arrayOf(PropTypes.string),
    vote_average: PropTypes.number,
    poster: PropTypes.string,
    overview: PropTypes.string,
  })).isRequired,
  video: PropTypes.string.isRequired,
  handleVideo: PropTypes.func.isRequired,
  getDetails: PropTypes.func.isRequired,
};
