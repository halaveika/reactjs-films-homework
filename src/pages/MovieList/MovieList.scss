.movieList-container {
  display: flex;
  height: auto;
  flex-direction: column;
  flex-wrap: nowrap;
  position: relative;
  width: 100vw;
  justify-content: center;
  align-content: center;
}

.active {
  top: 70px;
}

.movieItem-container {
  max-width: 1200px;
  width: 100%;
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  justify-content: space-around;
  padding: 0 30px 30px 30px;
  min-height: 70vh;
  justify-self: center;
  margin: 0 auto;
}

.noresult-msg {
  display: block;
  background: #1890ff;
  color: #fff;
  padding: 3px 5px;
  border-radius: 25%;
}

.pagination {
  width: 100%;
  display: flex;
  justify-content: center;
}
