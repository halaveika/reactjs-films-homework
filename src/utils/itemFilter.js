const itemfilter = (item, genre) => item
  .filter((e) => ((genre) ? e.genres.filter((name) => name === genre).length > 0 : e));

export default itemfilter;
