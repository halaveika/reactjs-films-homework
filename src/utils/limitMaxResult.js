import { MAX_SEARCH_RESULT, MAX_PAGE } from '../constants';

const limitMaxResult = (result) => {
  if (result.total_pages > MAX_PAGE) { result.total_pages = MAX_PAGE; }
  if (result.total_results > MAX_SEARCH_RESULT) { result.total_results = MAX_SEARCH_RESULT; }
  return result;
};

export default limitMaxResult;
