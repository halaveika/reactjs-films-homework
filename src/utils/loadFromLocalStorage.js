const loadFromLocalStorage = () => {
  const serializedState = localStorage.getItem('state');
  if (serializedState === null) return undefined;
  return JSON.parse(serializedState);
};

export default loadFromLocalStorage;
