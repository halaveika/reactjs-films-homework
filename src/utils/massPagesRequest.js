import {API_PAGE_SIZE} from '../constants';

const massPagesRequest = async(page, pageSize, httpService, str='') => {
  console.log('page: ' + page);
  console.log('pageSize: ' + pageSize);
  console.log('str: ' + str);
  const countPages = Math.ceil(pageSize / API_PAGE_SIZE);
  console.log('countPages: ' + countPages);
  let json;
  if(str){
    json = await Promise.all(Array(countPages).fill(1).map(async(_, index) =>  await httpService(str, page + index)));
  } else {
    json = await Promise.all(Array(countPages).fill(1).map(async(_, index) =>  await httpService(page + index)));
  }
  return {
    page: page,
    results: json.reduce((arr,e) => arr = arr.concat(e.results),[]),
    total_pages: json[json.length - 1].total_pages,
    total_results: json[json.length - 1].total_results,
  }

}

export default massPagesRequest;
