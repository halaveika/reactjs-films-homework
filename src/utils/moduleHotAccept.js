module.exports = {
  moduleHotAccept(mod) {
    if (mod && mod.hot) {
      mod.hot.accept();
    }
  },
};
