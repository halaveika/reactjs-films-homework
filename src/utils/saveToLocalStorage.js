const saveToLocalStorage = (state) =>{
    const serializedState = JSON.stringify(state);
    localStorage.setItem('state', serializedState);
  }

  export default saveToLocalStorage
