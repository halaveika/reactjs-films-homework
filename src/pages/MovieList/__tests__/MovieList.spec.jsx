import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import TestRenderer from 'react-test-renderer'; 
import '../../../utils/matchMedia';
import MovieList from '..';

const mockMovieListProps = {
  items: [{
    id: 2412412,
    title: 'Fight club',
    genres: ['Drama', 'Sport', 'Comedy'],
    vote_average: 8,
    poster: 'image_path',
    overview: 'Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry',
  },
  {
    id: 41412,
    title: 'Sopranos',
    genres: ['Drama', 'Criminal', 'comedy'],
    vote_average: 10,
    poster: 'image_path',
    overview: 'Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry',
  }],
  genres: [{
    id: 13,
    name: 'Adventure',
  }, {
    id: 18,
    name: 'Drama',
  }, {
    id: 20,
    name: 'Family',
  }, {
    id: 21,
    name: 'Fantasy',
  }],
  isLoading: false,
  filter: 'TRENDING',
  page: 1,
  total_results: 2,
  total_pages: 1,
  searchValue: '',
  isRow: true,
  genre: 'Drama',
  pageSize: 20,
  handleVideo: jest.fn(),
  getDetails: jest.fn(),
  getContent: jest.fn(),
  GetGenres: jest.fn(),
  setFilter: jest.fn(),
  setCurrentPage: jest.fn(),
  setGenre: jest.fn(),
  setListFlexDirection: jest.fn()
};

describe('test MovieList component', () => {

  let useEffect;
  const mockUseEffect = () => {
    useEffect.mockImplementationOnce(f => f());
  }
  
  beforeEach(()=> {
    useEffect = jest.spyOn(React, "useEffect");
    mockUseEffect();
    mockUseEffect();
  });

  it('should render MovieList component loaded', async() => {
    const renderer = new ShallowRenderer();
    let result;
    await TestRenderer.act(async() => {
      result = renderer.render(
        <MovieList
          items={mockMovieListProps.items}
          video={mockMovieListProps.video}
          genres={mockMovieListProps.genres}
          isLoading={mockMovieListProps.isLoading}
          filter={mockMovieListProps.filter}
          page={mockMovieListProps.page}
          total_pages={mockMovieListProps.total_pages}
          searchValue={mockMovieListProps.searchValue}
          isRow={mockMovieListProps.isRow}
          genre={mockMovieListProps.genre}
          handleVideo={mockMovieListProps.handleVideo}
          getDetails={mockMovieListProps.getDetails}
          getContent={mockMovieListProps.getContent}
          GetGenres={mockMovieListProps.GetGenres}
          setFilter={mockMovieListProps.setFilter}
          setCurrentPage={mockMovieListProps.setCurrentPage}
          setGenre={mockMovieListProps.setGenre}
          setListFlexDirection={mockMovieListProps.setListFlexDirection}
          total_results={mockMovieListProps.total_results}
          pageSize = {mockMovieListProps.pageSize}
        />,
      );
    });
    expect(result).toMatchSnapshot();
  });

  it('should render MovieList component loading', async() => {
    const renderer = new ShallowRenderer();
    let result;
    await TestRenderer.act(async() => {
      result = renderer.render(
        <MovieList
          items={mockMovieListProps.items}
          video={mockMovieListProps.video}
          genres={mockMovieListProps.genres}
          isLoading={true}
          filter={mockMovieListProps.filter}
          page={mockMovieListProps.page}
          total_pages={mockMovieListProps.total_pages}
          searchValue={mockMovieListProps.searchValue}
          isRow={mockMovieListProps.isRow}
          genre={mockMovieListProps.genre}
          handleVideo={mockMovieListProps.handleVideo}
          getDetails={mockMovieListProps.getDetails}
          getContent={mockMovieListProps.getContent}
          GetGenres={mockMovieListProps.GetGenres}
          setFilter={mockMovieListProps.setFilter}
          setCurrentPage={mockMovieListProps.setCurrentPage}
          setGenre={mockMovieListProps.setGenre}
          setListFlexDirection={mockMovieListProps.setListFlexDirection}
          total_results={mockMovieListProps.total_results}
          pageSize = {mockMovieListProps.pageSize}
        />,
      );
    });
    expect(result).toMatchSnapshot();
  });

  it('should render MovieList component in Column oder', async() => {
    const renderer = new ShallowRenderer();
    let result;
    await TestRenderer.act(async() => {
      result = renderer.render(
        <MovieList
          items={mockMovieListProps.items}
          video={mockMovieListProps.video}
          genres={mockMovieListProps.genres}
          isLoading={true}
          filter={mockMovieListProps.filter}
          page={mockMovieListProps.page}
          total_pages={mockMovieListProps.total_pages}
          searchValue={mockMovieListProps.searchValue}
          isRow={false}
          genre={mockMovieListProps.genre}
          handleVideo={mockMovieListProps.handleVideo}
          getDetails={mockMovieListProps.getDetails}
          getContent={mockMovieListProps.getContent}
          GetGenres={mockMovieListProps.GetGenres}
          setFilter={mockMovieListProps.setFilter}
          setCurrentPage={mockMovieListProps.setCurrentPage}
          setGenre={mockMovieListProps.setGenre}
          setListFlexDirection={mockMovieListProps.setListFlexDirection}
          total_results={mockMovieListProps.total_results}
          pageSize = {mockMovieListProps.pageSize}
        />,
      );
    });
    expect(result).toMatchSnapshot();
  });

});
