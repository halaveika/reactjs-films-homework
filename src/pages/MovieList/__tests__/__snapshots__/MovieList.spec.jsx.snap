// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test MovieList component should render MovieList component 1`] = `
<Layout
  className="movieList-container"
>
  <MovieItem
    genres={
      Array [
        "drama",
        "sport",
        "comedy",
      ]
    }
    getDetails={[MockFunction]}
    handleVideo={[MockFunction]}
    id={2412412}
    overview="Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry"
    poster="image_path"
    title="Fight club"
    video="youtube-link"
    vote_average={8}
  />
  <MovieItem
    genres={
      Array [
        "drama",
        "Criminal",
        "comedy",
      ]
    }
    getDetails={[MockFunction]}
    handleVideo={[MockFunction]}
    id={41412}
    overview="Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry"
    poster="image_path"
    title="Sopranos"
    video="youtube-link"
    vote_average={10}
  />
</Layout>
`;
